File: profitAndLoss/Portfolio.cpp
#include "Portfolio.hpp"


void Portfolio::initialisation(double optionPrice, const PnlVect *deltas, const PnlVect *prices)
{
	if (deltas_->size != deltas->size)
		pnl_vect_resize(deltas_, deltas->size);
	pnl_vect_clone(deltas_, deltas);
	risk_free_investment_ = optionPrice - pnl_vect_scalar_prod(deltas, prices);
}

void Portfolio::rebalancing(const PnlVect *deltas, const PnlVect *prices, double interest_rate, double time_step)
{
	pnl_vect_minus_vect(deltas_, deltas);
	risk_free_investment_ = risk_free_investment_ * exp(interest_rate * time_step) + pnl_vect_scalar_prod(deltas_, prices);
	pnl_vect_clone(deltas_, deltas);
}

double Portfolio::compute_value(const PnlVect *prices)
{
	return risk_free_investment_ + pnl_vect_scalar_prod(deltas_, prices);
}

double Portfolio::compute_final_value(const PnlVect *prices, double interest_rate, double time_step)
{
	return risk_free_investment_ * exp(interest_rate * time_step) + pnl_vect_scalar_prod(deltas_, prices);
}

File: EurostralWebApplication/Models/Portfolio.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace EurostralWebApplication.Models
{
    public class Portfolio
    {
        private int AssetNumber;

        private double LastRebalancing;

        public double[] Composition { get; set; }
        public double RiskFreeRateInvestment { get; set; }
        public double Value;

        public Portfolio(int assetNumber)
        {
            AssetNumber = assetNumber;
            Composition = new double[assetNumber];
            RiskFreeRateInvestment = 0;
            LastRebalancing = 0;
            Value = 0;
        }

        public void initialisation(double optionPrice, double[] composition, double[] prices, double currentTime)
        {
            Array.Copy(composition, Composition, AssetNumber);
            RiskFreeRateInvestment = optionPrice - Enumerable.Range(0, AssetNumber).Sum(i => Composition[i] * prices[i]);
            LastRebalancing = currentTime;
            Value = optionPrice;
        }

        public void rebalancing(double[] newComposition, double[] prices, double interestRate, double currentTime)
        {
            Composition = Composition.Zip(newComposition, (first, second) => first - second).ToArray();
            RiskFreeRateInvestment = RiskFreeRateInvestment * Math.Exp(interestRate * (currentTime - LastRebalancing)) + Enumerable.Range(0, AssetNumber).Sum(i => Composition[i] * prices[i]);
            Array.Copy(newComposition, Composition, AssetNumber);
            LastRebalancing = currentTime;
            Value = computeValue(prices);
        }

	    public double computeValue(double[] prices)
        {
            Value = RiskFreeRateInvestment + Enumerable.Range(0, AssetNumber).Sum(i => Composition[i] * prices[i]);
            return Value;
        }

        public double computeFinalValue(double[] prices, double interestRate, double timeStep)
        {
            Value = RiskFreeRateInvestment * Math.Exp(interestRate * timeStep) + Enumerable.Range(0, AssetNumber).Sum(i => Composition[i] * prices[i]);
            return Value;
        }
    }
}